File: app/src/main/java/com/academiaexpress/Fragments/LunchFragment.java
package com.academiaexpress.Fragments;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Point;
import android.os.Bundle;
import android.os.Handler;
import android.text.Html;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import com.academiaexpress.Activities.ProductsActivity;
import com.academiaexpress.Custom.ExpandableHeightGridView;
import com.academiaexpress.Data.DeliveryMeal;
import com.academiaexpress.Data.DeliveryOrder;
import com.academiaexpress.R;
import com.academiaexpress.Utils.Image;
import com.crashlytics.android.answers.Answers;
import com.crashlytics.android.answers.CustomEvent;

public class LunchFragment extends BaseProductFragment {

    private DeliveryMeal meal;
    private DeliveryOrder.OrderPart part;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_lunch, container, false);
    }

    public void animateScroll() {
        Point size = new Point();
        getActivity().getWindowManager().getDefaultDisplay().getSize(size);
        final int screenHeight = size.y;
        ((ScrollView) getView().findViewById(R.id.scroll)).smoothScrollTo(0,
                (screenHeight - getStatusBarHeight()) / 5);

        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                ((ScrollView) getView().findViewById(R.id.scroll)).smoothScrollTo((screenHeight - getStatusBarHeight()) / 5,
                        0);
            }
        }, 500);
    }

    @Override
    public void onStart() {
        super.onStart();

        if(!meal.isCanBuy()) {
            getView().findViewById(R.id.textViewfdfdfsfs).setVisibility(View.GONE);
        } else {
            getView().findViewById(R.id.textViewfdfdfsfs).setVisibility(View.VISIBLE);
        }

        getView().findViewById(R.id.textViewfdfdfsfs).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

            }
        });

        if(meal.getEnergy() == null) {
            getView().findViewById(R.id.en_name).setVisibility(View.GONE);
            getView().findViewById(R.id.en_content).setVisibility(View.GONE);
        } else {
            getView().findViewById(R.id.en_name).setVisibility(View.VISIBLE);
            getView().findViewById(R.id.en_content).setVisibility(View.VISIBLE);

            String[] energy = meal.getEnergy().split("energy");

            try {

                ((TextView) getView().findViewById(R.id.textView52a)).setText(energy[0]);
                ((TextView) getView().findViewById(R.id.textView53a)).setText(energy[1]);
                ((TextView) getView().findViewById(R.id.textView54a)).setText(energy[2]);
                ((TextView) getView().findViewById(R.id.textView55a)).setText(energy[3]);
            } catch (Exception e) {

            }
        }

        ((TextView) getView().findViewById(R.id.textView9)).setText(meal.getMealName());
        ((TextView) getView().findViewById(R.id.textView10)).setText(meal.getIngridients());
        ((TextView) getView().findViewById(R.id.textView16)).setText(meal.getDescription().equals("null") ? ""
                : meal.getDescription());
        ((TextView) getView().findViewById(R.id.textView11)).setText(Integer.toString(meal.getPrice()) + Html.fromHtml("<html>&#x20bd</html>").toString());
        Image.loadDishPhoto(meal.getPhotoLink(), ((ImageView) getView().findViewById(R.id.imageView3)));

        Point size = new Point();
        getActivity().getWindowManager().getDefaultDisplay().getSize(size);
        int screenHeight = size.y;
        FrameLayout layout = (FrameLayout) getView().findViewById(R.id.main);
        LinearLayout.LayoutParams params = (LinearLayout.LayoutParams)layout.getLayoutParams();
        params.height = screenHeight - getStatusBarHeight();


        getView().findViewById(R.id.textView).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(!DishFragment.Companion.getAnswer()) {
                    Answers.getInstance().logCustom(new CustomEvent("Товар добавлен"));
                }
                DishFragment.Companion.setAnswer(true);
                ((TextView) getView().findViewById(R.id.textView)).setText("ЗАКАЗАТЬ ЕЩЕ");
                ((ProductsActivity) getActivity()).addProduct(part);
            }
        });


        BaseAdapter adapter2 = new BaseAdapter() {
            @Override
            public int getCount() {
                return meal.getBadges().size();
            }

            @Override
            public Object getItem(int position) {
                return meal.getBadges();
            }

            @Override
            public long getItemId(int position) {
                return position;
            }

            @Override
            public View getView(int position, View convertView, ViewGroup parent) {
                convertView = LayoutInflater.from(getContext()).inflate(R.layout.item_ingredient, null);
                TextView textview = (TextView) convertView.findViewById(R.id.textView17);
                textview.setText(meal.getBadges().get(position).second);

                ImageView textview2 = (ImageView) convertView.findViewById(R.id.imageView6);
                textview2.setPadding((int)convertDpToPixel(25f, textview.getContext()),
                        (int)convertDpToPixel(25f, textview.getContext()),
                        (int)convertDpToPixel(25f, textview.getContext()),
                        (int)convertDpToPixel(25f, textview.getContext()));
                Image.loadPhoto(meal.getBadges().get(position).first, textview2);

                return convertView;

            }
        };

        ExpandableHeightGridView view2 = (ExpandableHeightGridView) getView().findViewById(R.id.gridView2);
        view2.setExpanded(true);
        view2.setAdapter(adapter2);
        adapter2.notifyDataSetChanged();

        ((LinearLayout) getView().findViewById(R.id.ingr)).removeAllViews();
        if(meal.getIngridientsList().size() == 0) {
            getView().findViewById(R.id.ingr).setVisibility(View.GONE);
        } else {
            getView().findViewById(R.id.ingr).setVisibility(View.VISIBLE);
            for (int i = 0; i < meal.getIngridientsList().size(); i++) {
                View v = LayoutInflater.from(getContext()).inflate(R.layout.lunch_part_card, null);

                ((TextView) v.findViewById(R.id.textView43)).setText(meal.getIngridientsList().get(i).first);
                ((TextView) v.findViewById(R.id.textView44)).setText(meal.getIngridientsList().get(i).second + "г");

                ((LinearLayout) getView().findViewById(R.id.ingr)).addView(v);
            }
        }
    }

    public ScrollView getScrollView() {
        return (ScrollView) getView().findViewById(R.id.scroll);
    }

    public int getStatusBarHeight() {
        int result = 0;
        int resourceId = getResources().getIdentifier("status_bar_height", "dimen", "android");
        if (resourceId > 0) {
            result = getResources().getDimensionPixelSize(resourceId);
        }
        return result; // + (int) convertDpToPixel(56f, getContext());
    }

    public void setPart(DeliveryOrder.OrderPart part) {
        this.part = part;
    }

    public void setInfo(DeliveryMeal meal) {
        this.meal = meal;
    }

    public static float convertDpToPixel(float dp, Context context) {
        Resources resources = context.getResources();
        DisplayMetrics metrics = resources.getDisplayMetrics();
        return dp * (metrics.densityDpi / 160f);
    }
}

File: app/src/main/java/com/academiaexpress/Fragments/LunchFragment.kt
package com.academiaexpress.Fragments

import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.academiaexpress.Activities.ProductsActivity
import com.academiaexpress.Base.BaseProductFragment
import com.academiaexpress.Custom.ExpandableHeightGridView
import com.academiaexpress.Data.Dish
import com.academiaexpress.Data.Order
import com.academiaexpress.R
import com.academiaexpress.Utils.AndroidUtilities
import com.academiaexpress.Utils.Image
import com.academiaexpress.Utils.LogUtil
import com.crashlytics.android.answers.Answers
import com.crashlytics.android.answers.CustomEvent
import java.util.*

class LunchFragment : BaseProductFragment() {
    private val ANIMATION_DURATION = 400L

    private lateinit var dish: Dish
    private lateinit var part: Order.OrderPart

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater!!.inflate(R.layout.fragment_lunch, container, false)
    }

    override fun animateScroll() {
        val coordinate = (AndroidUtilities.getScreenHeight(activity) - AndroidUtilities.getStatusBarHeight(context)) / 5
        Handler().post({ smoothScroll(0, coordinate) })
        Handler().postDelayed({ smoothScroll(coordinate, 0) }, ANIMATION_DURATION)
    }

    private fun smoothScroll(start: Int,  end: Int) {
        (view?.findViewById(R.id.scroll) as ScrollView?)?.smoothScrollTo(start, end)
    }

    private fun setInfo() {
        (view?.findViewById(R.id.name) as TextView?)?.text = dish.mealName
        (view?.findViewById(R.id.description) as TextView?)?.text = if (dish.ingredients == null) "" else dish.ingredients

        if (dish.description!!.isEmpty()) {
            (view?.findViewById(R.id.content) as TextView?)?.text = dish.description
        } else {
            view?.findViewById(R.id.about)?.visibility = View.GONE
        }

        (view?.findViewById(R.id.price) as TextView?)?.text = Integer.toString(dish.price) + getString(R.string.ruble_sign)

        Image.loadDishPhoto(dish.photoLink, view?.findViewById(R.id.image) as ImageView?)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val outOfStockIndicator = view?.findViewById(R.id.outOfStockIndicator)
        outOfStockIndicator?.visibility = if (dish.isOutOfStock) View.VISIBLE else View.GONE
        outOfStockIndicator?.setOnClickListener(null)
    }

    private fun hideEnergy() {
        view?.findViewById(R.id.en_content)?.visibility = View.GONE
    }

    private fun showEnergy() {
        view?.findViewById(R.id.en_content)?.visibility = View.VISIBLE
    }

    private fun setEnergy(energy: Array<String>) {
        (view?.findViewById(R.id.proteins) as TextView?)?.text = energy[0]
        (view?.findViewById(R.id.fats) as TextView?)?.text = energy[1]
        (view?.findViewById(R.id.carbohydrates) as TextView?)?.text = energy[2]
        (view?.findViewById(R.id.nutritional_value) as TextView?)?.text = energy[3]
    }

    private fun processEnergy() {
        if (dish.energy == null) {
            return
        }

        val energy = dish.energy!!.split("energy".toRegex()).dropLastWhile(String::isEmpty).toTypedArray()

        try {
            setEnergy(energy)
        } catch (e: Exception) {
            hideEnergy()
            LogUtil.logException(e)
        }
    }

    private fun parseEnergy() {
        if (dish.energy == null) {
            hideEnergy()
        } else {
            showEnergy()
            processEnergy()
        }
    }

    override fun onStart() {
        super.onStart()

        setInfo()
        parseEnergy()
        initMainView()
        setupAdapter()
        initIngredientsView()
    }

    private fun initIngredientsView() {
        if (dish.ingredientsList == null) {
            return
        }

        (view?.findViewById(R.id.ingr) as LinearLayout?)?.removeAllViews()
        if (dish.ingredientsList!!.size == 0) {
            view?.findViewById(R.id.ingr)?.visibility = View.GONE
        } else {
            showIngredientsView()
        }
    }

    private fun showIngredientsView() {
        if (dish.ingredientsList == null) {
            return
        }

        view?.findViewById(R.id.ingr)?.visibility = View.VISIBLE
        for (i in 0..dish.ingredientsList!!.size - 1) {
            val view = LayoutInflater.from(context).inflate(R.layout.item_lunch_part, null)

            (view.findViewById(R.id.name) as TextView).text = dish.ingredientsList!![i].first
            (view.findViewById(R.id.price) as TextView).text = dish.ingredientsList!![i].second + getString(R.string.ingredient_postfix)

            (view?.findViewById(R.id.ingr) as LinearLayout?)?.addView(view)
        }
    }

    private fun initMainView() {
        val layout = view?.findViewById(R.id.main) as FrameLayout?
        val params = layout?.layoutParams as LinearLayout.LayoutParams?
        params?.height = AndroidUtilities.getScreenHeight(activity) - AndroidUtilities.getStatusBarHeight(context)

        view?.findViewById(R.id.make_order_btn)?.setOnClickListener {
            if (!DishFragment.answer) {
                Answers.getInstance().logCustom(CustomEvent(getString(R.string.event_product_added)))
            }
            DishFragment.answer = true
            (view?.findViewById(R.id.make_order_btn) as TextView?)?.text = getString(R.string.order_more_code)
            (activity as ProductsActivity).addProduct(part)
        }
    }

    private fun setupAdapter() {
        val adapter = object : BaseAdapter() {
            override fun getCount(): Int {
                if (dish.badges == null) {
                    return 0
                }
                return dish.badges!!.size
            }

            override fun getItem(position: Int): Any {
                if (dish.badges == null) {
                    return ArrayList<String>()
                }
                return dish.badges!!
            }

            override fun getItemId(position: Int): Long { return position.toLong() }

            override fun getView(position: Int, convertView: View, parent: ViewGroup): View {
                var convertView = convertView
                convertView = LayoutInflater.from(context).inflate(R.layout.item_ingredient, null)

                if (dish!!.badges == null) {
                    return convertView
                }

                val name = convertView.findViewById(R.id.name) as TextView
                name.text = dish.badges!![position].second

                val icon = convertView.findViewById(R.id.image) as ImageView
                val padding = AndroidUtilities.dpToPx(icon.context, 25f)

                icon.setPadding(padding, padding, padding, padding)
                Image.loadPhoto(dish.badges!![position].first, icon)

                return convertView

            }
        }

        val gridView = view?.findViewById(R.id.gridView) as ExpandableHeightGridView?
        gridView?.isExpanded = true
        gridView?.adapter = adapter
        adapter.notifyDataSetChanged()
    }

    override fun getScrollView(): ScrollView? {
        return view!!.findViewById(R.id.scroll) as ScrollView?
    }

    override fun setPart(part: Order.OrderPart) {
        this.part = part
    }

    override fun setInfo(dish: Dish) {
        this.dish = dish
    }
}
